### PR TITLE
Implement error handling on MVVM and MVP modules

### PR DESCRIPTION
**Pull request by erichflock**

### CODE REVIEW

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 1]
```diff
@@ -7,6 +7,7 @@ import FeedFeature
 
 protocol FeedLoadingView {
 	func display(_ viewModel: FeedLoadingViewModel)
+    func display(error: String)
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:10]: I recommend separating the `display(error: String)` method into a new `FeedErrorView` protocol in this case, as they can be implemented by different views and controllers.

Moreover, we may decide to present errors for other actions rather than just `loading errors`.

Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:10]: This method could pass an optinal error string: `func display(error: String?)`, so the Presenter can decide when to show/hide the message.

For example, in the `didStartLoadingFeed()` method, the Presenter can call  `loadingView.display(error: .none)` to hide the error.

This way, the ViewController doesn't need to decide when to hide the error. It's much easier to manage presentation state when all changes are centralized in the presenter.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 2]
```diff
@@ -27,12 +31,17 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 	
 	func display(_ viewModel: FeedLoadingViewModel) {
 		if viewModel.isLoading {
+            errorView?.hideMessage()
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:34]: The Presenter should decide when to show/hide messages. See my other comment.

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 3]
```diff
@@ -280,6 +280,19 @@ final class FeedUIIntegrationTests: XCTestCase {
 		}
 		wait(for: [exp], timeout: 1.0)
 	}
+    
+    func test_loadFeedCompletion_rendersErrorMessageOnErrorUntilNextReload() {
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:284]: It's missing a test: should hide the error when the user taps on the error view.

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 4]
```diff
@@ -280,7 +280,20 @@ final class FeedUIIntegrationTests: XCTestCase {
 		}
 		wait(for: [exp], timeout: 1.0)
 	}
-	
+    
+    func test_loadFeedCompletion_rendersErrorMessageOnErrorUntilNextReload() {
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:284]: Same here - It's missing a test: should hide the error when the user taps on the error view.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 5]
```diff
@@ -17,17 +17,26 @@ final class FeedViewModel {
...
 	}
-
+    
+    var errorMessage: String {
+        Localized.Feed.errorMessage
+    }
+    
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+    var onFailure: Observer<Error>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:27]: `onFailure: Observer<Error> ` could be `onErrorStateChange: Observer<String?>`, so the ViewModel decides when to show/hide errors (instead of the view controller). It's much easier to manage view state when all changes are centralized in the ViewModel.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

Moreover, notice the type is optional string: `Observer<String?>` because the ViewModel should decide which formatted string to show in the UI. For example, if can call `onErrorStateChange(Localized.Feed.errorMessage)`

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 6]
```diff
@@ -28,10 +32,15 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		viewModel?.onLoadingStateChange = { [weak self] isLoading in
 			if isLoading {
 				self?.refreshControl?.beginRefreshing()
+                self?.errorView?.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:35]: The ViewModel should decide when to show/hide messages. See my other comment.

**Review by caiozullo — COMMENTED** [review]
Hi @erichflock,

You almost passed the challenge!

Please, check my review comments and let me know when you're done so I can review it again.

Also, please fix the PR name following the instructions in the readme file.

Thank you 🙂

**Comment by erichflock** [comment]
@caiozullo Thanks for the feedback! I have updated the PR with the changes.

**Comment by caiozullo** [comment]
Hi @erichflock,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

Great job!
